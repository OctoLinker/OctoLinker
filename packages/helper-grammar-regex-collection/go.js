import escapeRegexString from 'escape-regex-string';
import bylineParser from '@octolinker/helper-parser-byline';

const DOMAIN_REGEX = /^((?=[a-z0-9-]{1,63}\.)(xn--)?[a-z0-9]+(-[a-z0-9]+)*\.)+[a-z]{2,63}/;
const ALLOWED_DOMAINS = [
  'github.com',
  'bitbucket.org',
  'launchpad.net',
  'hub.jazz.net',
  'gopkg.in',
  'golang.org',
  'k8s.io',
];

// For go.mod file

function multiRequireRegExpBuilder(input) {
  return bylineParser(input, {
    openingPattern: /^require\s\($/,
    closingPattern: /^\)$/,
    matchPattern: /^\s?([\S]+)/,
  }).map(([value]) => {
    const domain = value.match(DOMAIN_REGEX);

    if (domain && !ALLOWED_DOMAINS.includes(domain[0])) {
      return;
    }

    const val = escapeRegexString(value);

    return new RegExp(`require\\s\\([^\)"]+(${val})`, 'gm');
  });
}

function singleRequireRegExpBuilder(input) {
  return bylineParser(input, {
    matchPattern: /require\s+([^\s\(\)"]+)/,
  }).map(([value]) => {
    const domain = value.match(DOMAIN_REGEX);

    if (domain && !ALLOWED_DOMAINS.includes(domain[0])) {
      return;
    }

    const val = escapeRegexString(value);

    return new RegExp(`require\\s+(${val})`, 'gm');
  });
}

// For regular go files

function multiImportRegExpBuilder(input) {
  return bylineParser(input, {
    openingPattern: /^import\s\($/,
    closingPattern: /^\)$/,
    matchPattern: /"([^"]+)"/,
  }).map(([value]) => {
    const domain = value.match(DOMAIN_REGEX);

    if (domain && !ALLOWED_DOMAINS.includes(domain[0])) {
      return;
    }

    const val = escapeRegexString(value);
    return new RegExp(`import\\s\\([^\)]+"(${val})"`, 'gm');
  });
}

function singleImportRegExpBuilder(input) {
  return bylineParser(input, {
    matchPattern: /import\s(?:[_\.]\s|[\w]+\s)?['"]([^'"\s]+)['"]?/,
  }).map(([value]) => {
    const domain = value.match(DOMAIN_REGEX);

    if (domain && !ALLOWED_DOMAINS.includes(domain[0])) {
      return;
    }

    const val = escapeRegexString(value);
    return new RegExp(
      `import\\s(?:[_\\.]\\s|[\\w]+\\s)?['"](${val})['"]?`,
      'gm',
    );
  });
}

export default function (blobSource) {
  return []
    .concat(
      multiImportRegExpBuilder(blobSource),
      singleImportRegExpBuilder(blobSource),
      multiRequireRegExpBuilder(blobSource),
      singleRequireRegExpBuilder(blobSource),
    )
    .filter((item) => !!item);
}
