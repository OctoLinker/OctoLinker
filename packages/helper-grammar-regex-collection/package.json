{
  "name": "@octolinker/helper-grammar-regex-collection",
  "version": "1.0.0",
  "description": "",
  "repository": "https://github.com/octolinker/octolinker/tree/main/packages/helper-grammar-regex-collection",
  "license": "MIT",
  "main": "./index.js",
  "dependencies": {
    "@octolinker/helper-parser-byline": "^1.0.0",
    "escape-regex-string": "^1.0.4",
    "xregexp": "^5.0.2"
  }
}
