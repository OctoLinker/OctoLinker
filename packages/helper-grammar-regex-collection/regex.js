import XRegExp from 'xregexp/lib/xregexp';
import build from 'xregexp/lib/addons/build';

build(XRegExp);

// This function makes it easy to build regular expressions that look more like formal grammars.
// See here for more info:
// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#Tagged_template_literals
// http://xregexp.com/api/#build
export default XRegExp.tag('xngm');
/* Usage Example:

    import regex from './regex';
    const sub = /inner/;
    const outer = regex`
      # comments start with #
      before

      # whitespace is insignificant, match it with \s
      \s

      # a reference to the 'sub' regex
      ${sub}

      \s

      # groups are not captured by default
      (after | after2)

      \s

      # this group is captured
      (?<name> after2 )
      $
    `;
    outer.exec('before inner after after2'); // ['before inner after after2', 'after2']
*/
