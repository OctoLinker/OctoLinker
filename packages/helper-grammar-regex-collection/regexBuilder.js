import XRegExp from 'xregexp/src/xregexp';
import build from 'xregexp/src/addons/build';

build(XRegExp);

// This function makes it easy to build regular expressions that look more like formal grammars.
// See here for more info:
// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#Tagged_template_literals
// http://xregexp.com/api/#build
export default function (flags = 'xngm') {
  const modeModifiers = flags.includes('x') ? '(?x)' : '';
  flags = Array.from(flags).filter(fl => fl !== 'x').join('');

  return (literals, ...substitutions) => {
    const subpatterns = {};

    const buildPattern = literals.raw.reduce((result, literal, index) => {
      const substitution = index in substitutions ? substitutions[index] : '';

      subpatterns[index] = substitution instanceof RegExp ? substitution :
        XRegExp.escape(substitution);

      return `${result + literal}{{${index}}}`;
    }, modeModifiers);

    return XRegExp.build(buildPattern, subpatterns, flags);
  };
}
/* Usage Example:

    import regexBuilder from './regexBuilder';
    const sub = /inner/;
    const xregexp = regexBuilder();
    const outer = xregexp`
      # comments start with #
      before

      # whitespace is insignificant, match it with \s
      \s

      # a reference to the 'sub' regex
      ${sub}

      \s

      # groups are not captured by default
      (after | after2)

      \s

      # this group is captured
      (?<name> after2 )
      $
    `;
    outer.exec('before inner after after2'); // ['before inner after after2', 'after2']
*/
