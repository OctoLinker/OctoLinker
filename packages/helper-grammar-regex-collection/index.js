import regex from './regex';

export { default as go } from './go.js';

const captureQuotedWord = regex`
  ['"]              # beginning quote
  (?<$1>[^'"\s]+)   # capture the word inside the quotes
  ['"]              # end quote
`;

const captureSpacedQuotedWord = regex`
  ['"]              # beginning quote
  (?<$1>[^'"]+)     # capture the word inside the quotes
  ['"]              # end quote
`;

const captureJsQuotedWord = regex`
  ['"\`]            # beginning quote
  (?<$1>[^'"\`\s]+) # capture the word inside the quotes
  ['"\`]            # end quote
`;

const diffSigns = regex`
  ^[ \t]*[+-]?
`;

const importMembers = regex`[\r\n\s\w{},*\$]*`;
const from = regex`\s from \s`;

export const NODEJS_RELATIVE_PATH = regex`
  __dirname
  \s* \+ \s*
  ${captureJsQuotedWord}
`;

export const NODEJS_RELATIVE_PATH_JOIN = regex`
  path\.join \s*
  \( \s*
  __dirname \s*
  , \s*
  ${captureJsQuotedWord}
`;

export const REQUIRE = regex`
  ( require(\.resolve)? | proxyquire | import | require_relative )
  \s* ( \s | \( ) \s*
  ${captureJsQuotedWord}
`;

export const IMPORT = regex`
  import \s ${importMembers}
  ${from}?
  ${captureQuotedWord}
`;

export const EXPORT = regex`
  export \s ${importMembers}
  ${from}
  ${captureQuotedWord}
`;

export const GEM = regex`
  gem \s ${captureQuotedWord}
`;

export const HOMEBREW = regex`
  (depends_on|conflicts_with)
  ( \s cask: | \s formula: )?
  \s
  ${captureQuotedWord}
`;

export const TYPESCRIPT_REFERENCE = regex`
  \/{3} \s?
  <reference \s path=${captureQuotedWord}
`;

export const DOCKER_FROM = regex`
  FROM \s+ (?<$1>[^\n]*)
`;

export const DOCKER_ENTRYPOINT = regex`
  ENTRYPOINT \s \[${captureQuotedWord}\]
`;

export const VIM_PLUGIN = regex`
  ${diffSigns}
  (
    (
      (Neo)?
      Bundle
      (Lazy|Fetch)?
    )
    |
    Plug(in)?
  ) \s
  ${captureQuotedWord}
`;

export const RUST_CRATE = regex`
  ${diffSigns}
  (extern \s crate | use)
  \s
  (?<$1>[^:;\s]+)
`;

export const PYTHON_IMPORT = regex`
  ${diffSigns}
  (import|from)
  \s
  (?<$1>[^\s]*)
  (\s import \s (?<$2>[^\s,]*))?
`;

export const REQUIREMENTS_TXT = regex`
  ${diffSigns}
  (?<$1>[\w-]+) # the package name
  (
    \s*
    [=~!><] # the beginning of a version specifier: https://www.python.org/dev/peps/pep-0440/#version-specifiers
    .*
  )? # the version is optional
  $
`;

export const HASKELL_IMPORT = regex`
  ${diffSigns}
  (import\s+(qualified\s)?)
  (?<$1> [A-Z][\w.]+)
`;

export const CSS_IMPORT = regex`
  ${diffSigns}
  @import
  \s
  ((url|URL)\()?
  ${captureQuotedWord}
`;

export const LESS_IMPORT = regex`
  ${diffSigns}
  @import
  \s
  ( \([a-z, ]+\) \s+ )? # http://lesscss.org/features/#import-options
  ((url|URL)\()?
  ${captureQuotedWord}
`;

export const HTML_IMPORT = regex`
  ${diffSigns}
  <link
  \s
  rel="import"
  \s
  href=${captureQuotedWord}>
`;

export const HTML_SCRIPT_IMPORT = regex`
  <script
  [^>]*
  \s
  src=${captureQuotedWord}
  [^>]*
  \s*>
`;

export const NET_PACKAGE = regex`
  <\s*
  package
  \s*
  id=${captureQuotedWord}
  .*/>
`;

export const GITHUB_ACTIONS = regex`
  uses:\s(?<$1>[^@\s]+)
`;

export const JAVA_IMPORT = regex`
  ${diffSigns}
  import
  \s
  (?<$1>
    (
      javax?
      |
      org\.springframework
      |
      io\.spring
      |
      sparklr\.common
      |
      org\.hamcrest
      |
      com\.fasterxml
      |
      org\.junit
      |
      org\.mockito
      |
      org\.apache
      |
      org\.slf4j
    )
    ((\.[a-zA-Z0-9_]+)+[a-zA-Z0-9_])*
  )
`;

export const NET_PROJ_PACKAGE = regex`
  <(DotNetCliTool|Framework|Package)?Reference
  \s+
  .*
  (Include|Update)=${captureQuotedWord}
  .*/?>
`;

export const NET_PROJ_SDK = regex`
  <Project
  \s+
  .*
  Sdk=${captureQuotedWord}
  .*>
`;

export const NET_PROJ_FILE_REFERENCE = regex`
  <(Compile|Content|EmbeddedResource|None|ProjectReference)
  \s+
  .*
  (Include|Update)=${captureSpacedQuotedWord}
  .*/?>
`;
