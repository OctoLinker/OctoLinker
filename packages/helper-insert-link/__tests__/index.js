import { REQUIRE } from '@octolinker/helper-grammar-regex-collection';
import insertLink from '../index';

describe('insert-link', () => {
  const DEFAULT_REGEX = /foo ("\w+")/g;
  const fakePlugin = {
    resolve: jest.fn().mockReturnValue('urlsToResolve'),
  };

  function helper(html, regex = DEFAULT_REGEX, plugin = fakePlugin, meta = {}) {
    const el = document.createElement('div');
    el.innerHTML = `<div id="LC1">${html}</div>`;

    const blob = {
      el,
      isDiff: false,
      firstLineNumber: 1,
      lineSelector() {
        return `#LC1`;
      },
      toString() {
        return el.textContent;
      },
    };

    const matches = insertLink(blob, regex, plugin, meta);

    // Remove root element for a pretty output
    const testEl = el.firstChild;
    testEl.removeAttribute('id');

    return {
      blob,
      el: testEl,
      matches,
    };
  }

  it('wraps the elements based on their char position which is specified in the keywords map', () => {
    expect(helper('foo <span><i>"</i>foo<i>"</i></span>').el).toMatchSnapshot();
  });

  it('wraps a nested element', () => {
    expect(
      helper('foo <div><span><i>"</i>foo<i>"</i></span></div>').el,
    ).toMatchSnapshot();
  });

  it('wraps double quotes', () => {
    const input = 'foo <span>"foo"</span>';

    expect(helper(input).el).toMatchSnapshot();
  });

  it('wraps single quotes', () => {
    const regex = /foo ('\w+')/g;
    const input = "foo <span>'foo'</span>";

    expect(helper(input, regex).el).toMatchSnapshot();
  });

  it('wraps mixed quotes', () => {
    const regex = /foo ('\w+")/g;
    const input = 'foo <span>\'foo"</span>';

    expect(helper(input, regex).el).toMatchSnapshot();
  });

  it('wraps a single word', () => {
    const regex = /foo (\w+)/g;
    const input = 'foo <span>bar</span>';

    expect(helper(input, regex).el).toMatchSnapshot();
  });

  it('wraps a single string', () => {
    const regex = /(bar)/g;
    const input = 'foo bar baz';

    expect(helper(input, regex).el).toMatchSnapshot();
  });

  it('wraps multiple strings', () => {
    const regex = /foo (bar)/g;
    const input = 'foo bar baz';

    expect(helper(input, regex).el).toMatchSnapshot();
  });

  it('wraps the element once', () => {
    const input = 'foo <span><i>"</i>foo<i>"</i></span>';
    const { blob, el, matches: match1 } = helper(input);
    const match2 = insertLink(blob, DEFAULT_REGEX, fakePlugin);

    expect(match1.length).toBe(1);
    expect(match2.length).toBe(0);

    expect(el).toMatchSnapshot();
  });

  it('does not remove closing parentheses from commented out require() calls', () => {
    const input = "// var faker = require('faker')";
    expect(helper(input, REQUIRE).el).toMatchSnapshot();
  });

  describe('returns', () => {
    it('returns an array', () => {
      const input = 'foo <span>"bar"</span>';

      expect(helper(input).matches.length).toBe(1);
      expect(helper(input).matches).toMatchSnapshot();
    });

    it('returns an array with values', () => {
      const input = 'foo <span>"bar"</span>';

      fakePlugin.resolve.mockReturnValue([undefined, null, '', 'bar']);
      expect(helper(input).matches[0].urls).toEqual(['bar']);
    });
  });

  describe('when findAndReplaceDOMText does not find any element', () => {
    it('wraps a nested element', () => {
      const regex = /foo (bar)/g;
      const input = 'foo b<span>ar</span>';

      expect(helper(input, regex).el).toMatchSnapshot();
    });

    it('wraps the matching group element', () => {
      const regex = /bar (bar)/g;
      const input = 'b<span>ar</span> b<span>ar</span>';

      expect(helper(input, regex).el).toMatchSnapshot();
    });
  });
});
