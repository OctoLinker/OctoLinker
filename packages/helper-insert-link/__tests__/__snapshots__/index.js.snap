// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`insert-link does not remove closing parentheses from commented out require() calls 1`] = `
<div>
  // var faker = require(
  <a
    class="octolinker-link"
    data-pjax="true"
  >
    'faker'
  </a>
  )
</div>
`;

exports[`insert-link returns returns an array 1`] = `
Array [
  Object {
    "link": <a
      class="octolinker-link"
      data-pjax="true"
    >
      "bar"
    </a>,
    "urls": "urlsToResolve",
  },
]
`;

exports[`insert-link when findAndReplaceDOMText does not find any element wraps a nested element 1`] = `
<div>
  foo 
  <a
    class="octolinker-link"
    data-pjax="true"
  >
    b
    <span>
      ar
      
    </span>
  </a>
</div>
`;

exports[`insert-link when findAndReplaceDOMText does not find any element wraps the matching group element 1`] = `
<div>
  
  b
  <span>
    ar
    
  </span>
   
  <a
    class="octolinker-link"
    data-pjax="true"
  >
    b
    <span>
      ar
      
    </span>
  </a>
</div>
`;

exports[`insert-link wraps a nested element 1`] = `
<div>
  foo 
  <div>
    <span>
      <i>
        
        "
      </i>
      <a
        class="octolinker-link"
        data-pjax="true"
      >
        foo
      </a>
      <i>
        "
        
      </i>
    </span>
  </div>
</div>
`;

exports[`insert-link wraps a single string 1`] = `
<div>
  foo 
  <a
    class="octolinker-link"
    data-pjax="true"
  >
    bar
  </a>
   baz
</div>
`;

exports[`insert-link wraps a single word 1`] = `
<div>
  foo 
  <span>
    <a
      class="octolinker-link"
      data-pjax="true"
    >
      bar
    </a>
  </span>
</div>
`;

exports[`insert-link wraps double quotes 1`] = `
<div>
  foo 
  <span>
    <a
      class="octolinker-link"
      data-pjax="true"
    >
      "foo"
    </a>
  </span>
</div>
`;

exports[`insert-link wraps mixed quotes 1`] = `
<div>
  foo 
  <span>
    <a
      class="octolinker-link"
      data-pjax="true"
    >
      'foo"
    </a>
  </span>
</div>
`;

exports[`insert-link wraps multiple strings 1`] = `
<div>
  foo 
  <a
    class="octolinker-link"
    data-pjax="true"
  >
    bar
  </a>
   baz
</div>
`;

exports[`insert-link wraps single quotes 1`] = `
<div>
  foo 
  <span>
    <a
      class="octolinker-link"
      data-pjax="true"
    >
      'foo'
    </a>
  </span>
</div>
`;

exports[`insert-link wraps the element once 1`] = `
<div>
  foo 
  <span>
    <i>
      
      "
    </i>
    <a
      class="octolinker-link"
      data-pjax="true"
    >
      foo
    </a>
    <i>
      "
      
    </i>
  </span>
</div>
`;

exports[`insert-link wraps the elements based on their char position which is specified in the keywords map 1`] = `
<div>
  foo 
  <span>
    <i>
      
      "
    </i>
    <a
      class="octolinker-link"
      data-pjax="true"
    >
      foo
    </a>
    <i>
      "
      
    </i>
  </span>
</div>
`;
